src/Test/Verilog/Expression.idr

```diff
@@ -0,0 +1,192 @@
+  SVMultArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Arr z)
+  SVDivBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVDivArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Arr z)
+  SVLessBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVLessArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Var Logic')
+  SVLessEqBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVLessEqArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Var Logic')
+  SVGreaterBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var z)
+  SVGreaterArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Var Logic')
+  SVLogicalEqualBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var Logic')
+  SVLogicalEqualArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Var Logic')
+  SVLogicalNotEqualBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var Logic')
+  SVLogicalNotEqualArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Var Logic')
+  SVCaseEqualBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var Bit')
+  SVCaseEqualArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Var Bit')
+  SVCaseNotEqualBits : SupremumType x y z => SVBinaryOperator (Var x) (Var y) (Var Bit')
+  SVCaseNotEqualArrays : ArrSupremum x y z => SVBinaryOperator (Arr x) (Arr y) (Var Bit')
+
+||| Predicate to test if the type is integral as described in the IEEE 1800-2023 standard.
+public export
+data IsIntegralType : SVType -> Type where
+  LogicIsIntegral : IsIntegralType (Var Logic')
+  BitIsIntegral : IsIntegralType (Var Bit')
+  WireIsIntegral : IsIntegralType (Var Wire')
+  UwireIsIntegral : IsIntegralType (Var Uwire')
+  IntIsIntegral : IsIntegralType (Var Int')
+  IntegerIsIntegral : IsIntegralType (Var Integer')
+  UnpackedArrysIsIntegral : IsIntegralType t -> IsIntegralType (Arr (Unpacked t s e))
+  PackedArrysIsIntegral : AllowedInPackedArr t => IsIntegralType t -> IsIntegralType (Arr (Packed t s e))
+
+public export
+data IsLogicOrBite : SVType -> Type where
+  LogicIs : IsLogicOrBite (Var Logic')
+  BitIs : IsLogicOrBite (Var Bit')
+  WireIs : IsLogicOrBite (Var Wire')
+  UwireIs : IsLogicOrBite (Var Uwire')
+
+public export
+data SVUnaryOperator : (iType : SVType) -> (oType : SVType) -> Type where
+  SVUnaryMinus : IsIntegralType iType => SVUnaryOperator iType iType
+  SVLogicalNegation : IsLogicOrBite iType => SVUnaryOperator iType iType
+  SVBitwiseNegationUnpacked : IsIntegralType iType => SVUnaryOperator (Arr $ Unpacked iType s e) (Arr $ Unpacked iType s e)
+  SVBitwiseNegationPacked : AllowedInPackedArr iType => IsIntegralType iType => SVUnaryOperator (Arr $ Packed iType s e) (Arr $ Packed iType s e)
+
+public export
+data SVExpression : (expressionType : SVType) -> (ports : PortsList) -> (usedPorts : ListOfPortsIndices ports) -> Type where
```

Comment:
Does this `ports` list represent connections outside the context of the expression?